astropy/units/quantity_helper/converters.py

```diff
@@ -338,10 +338,15 @@ def check_output(output, unit, inputs, function=None):
         # check we can handle the dtype (e.g., that we are not int
         # when float is required).  Note that we only do this for Quantity
         # output; for array output, we defer to numpy's default handling.
-        if not np.can_cast(np.result_type(*inputs), output.dtype,
-                           casting='same_kind'):
-            raise TypeError("Arguments cannot be cast safely to inplace "
-                            "output with dtype={}".format(output.dtype))
+        # Also, any structured dtype are ignored (likely erfa ufuncs).
+        # TODO: make more logical; is this necessary at all?
```

Comment:
Agreed. I'm having trouble following this.

Reply:
It is to safeguard against, say, an output with integer dtype being used for a float operation, say,
```
q = u.Quantity([1, 2], u.m, dtype=int)
q+=1*u.cm
# TypeError: Arguments cannot be cast safely to inplace output with dtype=int64
```
This is partially a leftover from when the default dtype was not float. Anyway, for later...